src/orion/core/worker/experiment.py

```diff
@@ -243,24 +246,31 @@ def reserve_trial(self, score_handle=None):
                         "parameter space has not been defined yet.")
 
         selected_trial = random.sample(new_trials, 1)[0]
-
-        # Query on status to ensure atomicity. If another process change the
-        # status meanwhile, read_and_write will fail, because query will fail.
-        query = {'_id': selected_trial.id, 'status': selected_trial.status}
+        # log.debug(f'{">" * _depth} selected (trial: {selected_trial})')
 
         update = dict(status='reserved', heartbeat=datetime.datetime.utcnow())
 
         if selected_trial.status == 'new':
             update["start_time"] = datetime.datetime.utcnow()
 
-        reserved = self._db.write('trials', query=query, data=update)
+        # Query on status to ensure atomicity. If another process change the
+        # status meanwhile, update will fail, because query will fail.
+        # This relies on the atomicity of document updates.
+        # /!\ This is not atomic if you are using monogo db replica set !
```

Comment:
Following our discussion I believe you can remove this comment. Since we only rely on writes to handle race-conditions, and writes are atomic on the primary set, then this cannot fail. The only issue is about reads on replica sets which may be incoherent, but that is fine because either the status is marked as reserved and then the trial simply isn't reserved at this moment, or it is marked as available (new, interrupted, ...) and the reservation will fail. Either case, we are fine.